webtasks/cron_mongodb.js

```diff
@@ -0,0 +1,230 @@
+var _ = require('lodash');
+var Bluebird = require('bluebird');
+
+var mongo;
+
+
+var webtask = function (context, req, res) {
+    var action = validate_method({
+        put_job: 'PUT',
+        get_job: 'GET',
+        destroy_job: 'DELETE',
+        list_jobs: 'GET',
+        job_history: 'GET',
+        reserve_jobs: 'POST',
+        renew_reservation: 'PUT',
+        release_reservation: 'DELETE',
+    });
+    
+    if (!validate_params(['MONGO_COLLECTION']))
+        return;
+    
+    if (action === 'put_job') {
+        // Create or update a scheduled webtask
+        
+        if (!validate_params(['container', 'name', 'schedule', 'token']))
+            return;
+            
+        var now = new Date();
+        
+        var item = {
+            state: 'active',
+            schedule: context.data.schedule,
+            token: context.data.token,
+            container: context.data.container,
+            name: context.data.name,
+            last_run_at: null,
+            next_available_at: now,
+            last_result: null,
+            run_count: 0,
+            error_count: 0,
+        };
+        
+        return withMongoCollection(context.data.MONGO_COLLECTION)
+            .then(function (coll) {
+                return coll.findOneAndUpdateAsync({
+                    container: item.container,
+                    name: item.name,
+                }, item, {
+                    returnOriginal: false,
+                    upsert: true,
+                });
+            })
+            .get('value')
+            .then(function (data) {
+                res.writeHead(200, {'Content-Type': 'application/json'});
+                res.end(JSON.stringify(data));
+            })
+            .catch(function (err) {
+                error(err.statusCode || 500, err);
+            });
+    } else if (action === 'reserve_jobs') {
+        // Reserve a set of jobs for execution
+        
+        if (!validate_params(['count', 'ttl']))
+            return;
+        
+        var now = new Date();
+        var next_available_at = new Date(now.valueOf() + (parseInt(context.data.ttl, 10) * 1000))
+        
+        return withMongoCollection(context.data.MONGO_COLLECTION)
+            .then(function (coll) {
+                return Bluebird.resolve(_.range(context.data.count))
+                    // TODO (ggoodman): Refactor this to use reduce to avoid calling N queries if reservation fails before N
```

Comment:
There are few important issues (imho) here:

1. Ensure that if one query fails, the successful reservations are passed on (see commit). I will need your help thinking of a way to test this.
2. Ensure that only successful queries--that actually did an update--are returned (this is why I filter on Boolean)
3. Stop doing queries if the queue is already exhausted (in cases of large N, this could introduced unnecessary latency). This is not yet addressed and was the reason for my TODO.